BCs (#134)
festim-dev/FESTIM-workshop: book/content/boundary_conditions/concentration.md

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.18.1
kernelspec:
  display_name: festim-workshop
  language: python
  name: python3
---

> # Concentration #

Reviewer: I think instead of Concentration + Particle Flux + Surface reactions, we can have everything under "Hydrogen Transport"?

Boundary conditions (BCs) are essential to FESTIM simulations, as they describe the mathematical problem at the boundaries of the simulated domain. Read more about BCs _[here](https://festim.readthedocs.io/en/fenicsx/userguide/boundary_conditions.html)_.

This tutorial goes over how to define concentration boundary conditions for hydrogen transport simulations.

Objectives:
* Define a fixed concentration BC
* Choose which solubility law (Sieverts' or Henry's)
* Solve a hydrogen transport problem with plasma implantation

+++

> ## Defining fixed concentration ##

Reviewer: This section needs some "interactivity" as right now it could simply be a section of the documentation. We learn how to define a FixedConcentrationBC object but not how to use it in a global problem.

How does it work if we want to add several BCs?
How can I define BCs for different species?
What happens if I don't define a BC at a boundary?

How about adding a full problem at the end of the section:
- steady state
- 2D
- top BC: c = 2 x + y
- right BC: c = y**2 +1

> BCs need to be assigned to surfaces using FESTIM's `SurfaceSubdomain` class. To define the concentration of a specific species, we can use `FixedConcentrationBC`:

Reviewer: ```suggestion
BCs need to be assigned to surfaces using FESTIM's `SurfaceSubdomain` class. To set the concentration of a specific species on a boundary, we can use `FixedConcentrationBC`:
```

```{code-cell} ipython3
from festim import FixedConcentrationBC, Species, SurfaceSubdomain

boundary = SurfaceSubdomain(id=1)
H = Species(name="Hydrogen")

my_bc = FixedConcentrationBC(subdomain=boundary, value=10, species=H)
```

The imposed concentration can be dependent on space, time and temperature, such as:

$$ 

> BC = 10 + x^2 + T(t+2)

Reviewer: ```suggestion
c = 10 + x^2 + T(t+2)
```

$$

> 

Reviewer: ```suggestion
To do so, we define a python function.
```

```{code-cell} ipython3

> my_custom_value = lambda x, t, T: 10 + x[0]**2 + T *(t + 2)

Reviewer: We should also note that (x[0], x[1], x[2]) corresponds to (x, y, z) in a cartesian space

my_bc = FixedConcentrationBC(subdomain=boundary, value=my_custom_value, species=H)

> ```

Reviewer: Here we should add a note block showing that this is equivalent to 

```python
def my_custom_value(x, t, T):
    return 10 + x[0]**2 + T *(t + 2)
```

> ## Choosing a solubility law ##

Reviewer: Same comment as above. We need to show more about the integration. For example, show that this is equivalent to a FixedConcentrationBC but it's just providing convenience

Users can define the surface concentration using either Sieverts’ law, $c = S(T)\sqrt P$, or Henry's law, $c=K_H P$, where $S(T)$ and $K_H$ denote the temperature-dependent Sieverts’ and Henry’s solubility coefficients, respectively, and $P$ is the partial pressure of the species on the surface. 

For Sieverts' law of solubility, we can use `festim.SievertsBC`:

```{code-cell} ipython3
from festim import SievertsBC, SurfaceSubdomain, Species

boundary = SurfaceSubdomain(id=1)
H = Species(name="Hydrogen")

custom_pressure_value = lambda t: 2 + t
my_bc = SievertsBC(subdomain=3, S_0=2, E_S=0.1, species=H, pressure=custom_pressure_value)
```

Similarly, for Henry's law of solubility, we can use `festim.HenrysBC`:

```{code-cell} ipython3
from festim import HenrysBC

pressure_value = lambda t: 5 * t
my_bc = HenrysBC(subdomain=3, H_0=1.5, E_H=0.2, species=H, pressure=pressure_value)
```

> ## Plasma implantation approximation ##

Reviewer: Maybe this one could go in a section "Hydrogen transport: advanced"

> ## Plasma implantation approximation ##

Reviewer: For this application, because it's a "real-world" application, I would use real numbers.

Typically, the implantation range is a few nanometers for example.

> ## Plasma implantation approximation ##

Reviewer: There's actually an issue here: this is not a boundary condition approximation, you set an actual volumetric source.

+++

We can also approximate plasma implantation using FESTIM's `ParticleSource` class, which is helpful in modeling thermo-desorption spectra (TDS) experiments or including the effect of plasma exposure on hydrogen transport. Learn more about how FESTIM approximates plasma implantation _[here](https://festim.readthedocs.io/en/fenicsx/theory.html)_.

Consider the following 1D plasma implantation problem, where we represent the plasma as a hydrogen source $S_{ext}$:

$$ S_{ext} = \varphi \cdot f(x) $$

> $$\varphi = 1\cdot 10^{13} \quad \mathrm{m}^{-2}\mathrm{s}^{-1}$$

Reviewer: ```suggestion
$$\varphi = 10^{13} \quad \mathrm{m}^{-2}\mathrm{s}^{-1}$$
```
